### PR TITLE
Feature: Add Map Repository, Markers, Area Geometry Logic, and Comprehensive Unit Tests

### LINKED ISSUE #144: Implement markers and areas logic
Implement core logic for map markers and polygonal areas, including geometry validation, point-in-polygon detection and in-memory repository support with comprehensive unit tests.

### PR DESCRIPTION
**Pull request by noa-flo**
## What has been changed?
This PR introduces a complete local map domain layer with in-memory storage, geometry validation, data classes for locations, markers, areas and robust unit testing.  

The main updates include:

### **Core Data Models**
- **`Location`**: Represents geographic coordinates (latitude, longitude) with an optional label.
- **`Marker`**: Represents a point on the map.  
  - Added a secondary constructor for convenience (latitude/longitude directly).  
  - Auto-generates a unique `id` via UUID.  
- **`Area`**: Represents a polygonal region defined by at least **three distinct markers**.  
  - Enforces geometric validity (`require >= 3 distinct markers`).  
  - Markers are automatically sorted counter-clockwise around the centroid using an ear-clipping–based algorithm.
  - Implements `contains(marker)` to test whether a marker lies inside the polygon, using the ray-casting algorithm.

### **Utility**
- **`GeoUtils`**:  
  - Added `isPointInPolygon()` function implementing the **even–odd ray-casting algorithm** for point-in-polygon detection.  
  - Added `sortMarkersCounterClockwise()` that order a list of markers around the centroid in counter-clockwise order and applies ear-clipping logic to handle complex and concave polygons.
  - Added helper functions `cross()` and `pointInTriangle()` for polygon geometry and ear-clipping triangulation.

### **Repository Implementation**
- **`MapRepository` interface**: Defines contract for managing `Marker` and `Area` entities.  
- **`MapRepositoryLocal`**: In-memory implementation using `mutableMapOf` for both markers and areas.  
  - Supports CRUD operations for markers and creation/retrieval of polygonal areas.  
  - Validates distinct marker constraints when creating new areas.

### **Unit Tests Added**
- **`LocationTest`**: Ensures correct field assignment and equality.  
- **`MarkerTest`**:  
  - Verifies secondary constructor sets location properly.  
  - Checks unique UUID generation.  
- **`AreaTest`**:  
  - Validates the polygon constraint (≥ 3 distinct markers).  
  - Tests `contains()` for points inside, outside, on edges, and on vertices.  
- **`GeoUtilsTest`**:  
  - Covers various geometric shapes (square, triangle, arrow).  
  - Includes corner cases like edge and vertex hits.  
- **`MapRepositoryLocalTest`**:  
  - Tests marker addition, removal, and retrieval.  
  - Ensures area creation enforces distinct marker requirement.  
  - Validates repository integrity with multiple areas and distinct marker IDs.

Close #144 (Task)
Related to #43 (User story)

### CODE REVIEW

File: app/src/test/java/com/android/sample/model/map/AreaTest.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+class AreaTest {
+
+  private lateinit var area: Area
+  private lateinit var insidePoint: Marker
+  private lateinit var outsidePoint: Marker
+
+  @Before
+  fun setup() {
+    // Define markers forming a square area
+    val m1 = Marker(latitude = 48.8566, longitude = 2.3522)
```
Comment (Poketimi) [inline app/src/test/java/com/android/sample/model/map/AreaTest.kt]: For consistency, couldn't we name those variable s1...s4?

File: app/src/main/java/com/android/sample/model/map/Marker.kt [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  /**
+   * Secondary constructor allowing to create a Marker directly from latitude and longitude without
+   * having to manually construct a Location object.
+   */
+  constructor(
+      id: String = UUID.randomUUID().toString(),
+      latitude: Double,
+      longitude: Double,
+      label: String? = null
+  ) : this(id = id, location = Location(latitude, longitude, label), label = label)
```
[the remark this answers is not in the document]
Comment (Poketimi) [reply app/src/main/java/com/android/sample/model/map/Marker.kt]: It is true that I do not really see the purpose to assign to time the same label. 
Additionally, I'm dubious about needing a label field to describe a location since we will I think only use them in marker.

That being said, not a big deal to do it that way =)

File: app/src/main/java/com/android/sample/model/map/Area.kt [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * Represents a polygonal area defined by a list of map markers.
+ *
+ * An Area groups multiple Marker instances to form a polygonal region on a map.
+ *
+ * @property id Unique identifier for the area. Defaults to a randomly generated UUID string.
+ * @property label Optional human-readable label or name for the area.
+ * @property markers List of Marker objects that define the vertices of the area polygon. The
+ *   polygon is expected to be formed in the order provided by this list.
```
[the remark this answers is not in the document]
Comment (noa-flo) [reply app/src/main/java/com/android/sample/model/map/Area.kt]: We need the markers to be in order because most polygon operations, like checking if a point is inside the polygon (contains), rely on traversing the vertices in the correct contour order. If the points are unordered, algorithms like ray-casting can produce incorrect results.

To remove the burden from users and guarantee correctness, the Area constructor will sorts the markers automatically (by their angular position around the polygon’s centroid) in a future commit.

File: app/src/main/java/com/android/sample/model/map/Location.kt [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ */
+data class Location(
+    /** Latitude in decimal degrees (usually -90.0..90.0). */
+    val latitude: Double,
+
+    /** Longitude in decimal degrees (usually -180.0..180.0). */
+    val longitude: Double,
+
+    /** Optional short label for display purposes. Null when no label is provided. */
+    val label: String? = null,
```
Comment (Qoookie) [inline app/src/main/java/com/android/sample/model/map/Location.kt]: I'm not sure to see the purpose of this label, since a Marker already take a label and a Marker's localisation is not null, we won't use location for his label, I think this can be safely remove !

File: app/src/test/java/com/android/sample/model/mapRepositoryTest/MapRepositoryLocalTest.kt [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    val all = repository.getAllMarkers()
+    assertEquals(4, all.size)
+  }
+
+  @Test
+  fun createArea_withLessThan3Markers_shouldThrowException() {
+    try {
+      repository.createArea(label = "InvalidArea", markerIds = listOf(marker1.id, marker2.id))
+      fail("Expected IllegalArgumentException for less than 3 distinct markers")
+    } catch (_: IllegalArgumentException) {}
```
Comment (Qoookie) [inline app/src/test/java/com/android/sample/model/mapRepositoryTest/MapRepositoryLocalTest.kt]: Very good test for a Futur proof code !

File: app/src/test/java/com/android/sample/model/map/GeoUtilsTest.kt
[the remark this answers is not in the document]
Comment (noa-flo) [reply app/src/test/java/com/android/sample/model/map/GeoUtilsTest.kt]: Indeed, this is a good remark, my previous implementation of sorting CCW the markers was incorrect in the case of concave polygones...

I updated the sorting to centroid + ear clipping, so concave shapes now work. I also added a test with an arrow-shaped polygon.

This is implemented in commit 5a95ce30.

**Comment by Poketimi** [comment]
I just saw that sonarcloud has an issue, is it possible to make sortedByAngle immutable in GeoUtils?

**Comment by noa-flo** [comment]
> I just saw that sonarcloud has an issue, is it possible to make sortedByAngle immutable in GeoUtils?

Yes, this is done in commit 12dd0b1.

**Review by Qoookie — APPROVED** [review]
This seems all good, and very well tested ! For me we can merge this !
